### PR TITLE
Route based ACLs POC

### PR DESCRIPTION
**Pull request by superkhau**
Connect to https://github.com/strongloop-internal/scrum-loopback/issues/378

### CODE REVIEW

File: server/middleware.json [hunk 1]
```diff
@@ -47,6 +47,14 @@
       }
     }
   },
+  "auth:after": {
+    "./middleware/routes-acl": {
+      "params": {
+        "hello": 1,
```
Comment (superkhau) [inline server/middleware.json]: @raymondfeng What kind of syntax should we use here?

File: server/middleware/routes-acl/index.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports = function(options) {
+  options = options || {};
+  console.log(options);
```
Comment (raymondfeng) [inline server/middleware/routes-acl/index.js]: It should return a middleware handler, for example.

return function checkRouteACLs(req, res, next) {...};

File: test/routes-acl.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+describe.only('Route based ACLs', function() {
+
+  before(require('./start-server'));
+
+  after(function(done) {
+    app.close(done);
+  });
+
+  it('should ...', function() {
```
Comment (superkhau) [inline test/routes-acl.js]: What does an example test look like? Should I be making a request to `/api/notes` and making sure the ACL is triggered?

File: server/middleware/routes-acl/index.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+module.exports = function(options) {
+  options = options || {};
+  return function checkRouteAcls(req, res, next) {
```
Comment (superkhau) [inline server/middleware/routes-acl/index.js]: @raymondfeng Done. Next?

File: server/middleware.json [hunk 5]
```diff
@@ -47,6 +47,32 @@
       }
     }
   },
+  "auth:after": {
+    "./middleware/routes-acl": {
+      "params": {
+        "rules": [
```
Comment (superkhau) [inline server/middleware.json]: something like this?

File: server/middleware/routes-acl/index.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    // determine which route is being requested
+    // loop through routes-acl rules to check for matches
+    // if match
+    //    check the req user id and app id for perms
+    //    if allowed do nothing (let request through)
+    //    if not allowed
+    //      return 403
+    // if no match
+    //    do nothing (let request through)
```
Comment (superkhau) [inline server/middleware/routes-acl/index.js]: @raymondfeng Does this pseudocode make sense to you?

File: test/routes-acl.js (hunk 3 above)
Comment (superkhau) [inline test/routes-acl.js]: @raymondfeng What does the request look like?

File: server/middleware.json [hunk 7]
```diff
@@ -47,6 +47,32 @@
...
+          {
+            "route": "/api/notes",
```
Comment (raymondfeng) [inline server/middleware.json]: The route should be able to specify `verb` as well.

File: server/middleware.json [hunk 8]
```diff
@@ -47,6 +47,32 @@
...
+              "accessType": "*",
+              "principalType": "ROLE",
+              "principalId": "$everyone",
+              "permission": "ALLOW"
+            }
+          }
+        ]
+      }
+    }
+  },
```
Comment (raymondfeng) [inline server/middleware.json]: I also want to have the ability to create `scopes` to group routes and the ACL should apply to `scopes`

Comment (raymondfeng) [inline server/middleware.json]: See some discussion at https://github.com/strongloop/loopback/issues/625

File: server/middleware/routes-acl/index.js (hunk 6 above)
Comment (raymondfeng) [inline server/middleware/routes-acl/index.js]: Sure. Ultimately, we should be able to just call the ACL engine that understands all the rules.

File: test/routes-acl.js (hunk 3 above)
Comment (raymondfeng) [inline test/routes-acl.js]: The request is usually an api call with the `access_token`. The oAuth2 resource server will establish the appId/userId from the access token. Such ids can then be used by the ACL.

**Comment by superkhau** [comment]
@raymondfeng Review please.

**Pull request closed without merging**